src/dodal/beamlines/beamline_utils.py

```diff
@@ -30,58 +36,87 @@ def list_active_devices() -> List[str]:
     define lists of devices in beamline files. Additional keyword arguments are passed
     directly to the device constructor.
 
     Arguments:
-        device: Callable        the device class
-        name: str               the name for ophyd
-        prefix: str             the PV prefix for the most (usually all) components
-        wait: bool              whether to run .wait_for_connection()
-        fake: bool              whether to fake with ophyd.sim
-        post_create: Callable   (optional) a function to be run on the device after
-                                creation
-        bl_prefix: bool         if true, add the beamline prefix when instantiating, if
-                                false the complete PV prefix must be supplied.
+        device_factory: Callable    the device class
+        name: str                   the name for ophyd
+        prefix: str                 the PV prefix for the most (usually all) components
+        wait: bool                  whether to run .wait_for_connection()
+        fake: bool                  whether to fake with ophyd.sim
+        post_create: Callable       (optional) a function to be run on the device after
+                                    creation
+        bl_prefix: bool             if true, add the beamline prefix when instantiating, if
+                                    false the complete PV prefix must be supplied.
     Returns:
         The instance of the device.
     """
-    active_device = ACTIVE_DEVICES.get(name)
+    device_instance: T
+    already_existing_device: Optional[AnyDevice] = ACTIVE_DEVICES.get(name)
     if fake:
-        device = make_fake_device(device)
-    if active_device is None:
-        ACTIVE_DEVICES[name] = device(
+        device_factory = make_fake_device(device_factory)
+    if already_existing_device is None:
+        device_instance = device_factory(
             name=name,
             prefix=f"{(BeamlinePrefix(BL).beamline_prefix)}{prefix}"
             if bl_prefix
             else prefix,
             **kwargs,
         )
+        ACTIVE_DEVICES[name] = device_instance
         if wait:
-            ACTIVE_DEVICES[name].wait_for_connection()
+            _wait_for_connection(device_instance, sim=fake)
+
     else:
-        if not active_device_is_same_type(active_device, device):
+        if not active_device_is_same_type(already_existing_device, device_factory):
             raise TypeError(
-                f"Can't instantiate device of type {type(active_device)} with the same "
+                f"Can't instantiate device of type {device_factory} with the same "
                 f"name as an existing device. Device name '{name}' already used for "
-                f"a(n) {device}."
+                f"a(n) {type(already_existing_device)}."
             )
+        else:
+            # We have manually checked types
+            device_instance = cast(T, ACTIVE_DEVICES[name])
```

Comment:
It doesn't pass type-checking without the cast - as `already_existing_device` is an `Optional[AnyDevice]`, type checker can't prove that it can be assigned to `device_instance` which is of type `T`.

I have removed the `else:` as suggested.